File: chains/rust/zkvm/src/vm.rs
// SPDX-License-Identifier: BSD-3-Clause-Eco
// Copyright (C) 2019-2026, Lux Industries Inc. All rights reserved.

//! The chain itself: what it holds, what it admits, and what it commits.
//!
//! ## One predicate
//!
//! [`Zvm::admit`] is the whole of what a transaction must satisfy, and both
//! sides ask it — assembly in [`Zvm::build`] and consensus in [`Zvm::verify`].
//! A proposer that assembled a transaction its own peers refuse produces a
//! block nobody accepts and, since nothing evicts it, never produces another.
//!
//! ## Two halves of one verification
//!
//! [`Zvm::verify`] is a single pass, in the reference's order: the block's own
//! shape, then each transaction's shape and its proof, then the parent and the
//! tip it must sit on, then the state root it commits to. There is no separate
//! syntactic method to call, so where the two halves divide is stated over the
//! RULES, in [`crate::error::Error::before_the_chain`], rather than over the
//! code that happens to hold them — three of those rules are block-level and
//! could not live in a per-transaction pass at all.
//!
//! ## Acceptance is one commit or none
//!
//! [`Zvm::accept`] stages the block record, its height entry, the tip pointer,
//! every nullifier it spends, every output it creates and the advanced state
//! root through ONE view, and commits them together. A chain that wrote as it
//! went would have a first failed write leave some notes spent and some outputs
//! created under a tip it had already moved — a shielded pool half applied,
//! with no way back and no way to apply the block again.

use crate::block::{self, Block};
use crate::config::Config;
use crate::db::{Memory, View};
use crate::error::{Error, Result};
use crate::ids::{self, Id};
use crate::nullifier::Nullifiers;
use crate::root::Root;
use crate::store::{Decided, Store};
use crate::txs::Tx;
use crate::utxo::{Utxo, Utxos};
use crate::verifier::Verifier;

/// What this chain calls itself.
pub const NAME: &str = "zkvm";
/// The version it reports.
pub const VERSION: &str = "1.0.0";

/// What a Z-chain is born with.
///
/// The timestamp is hashed into the genesis block id, so a genesis that names
/// none is stamped 0 rather than read off the wall clock — otherwise every node
/// derives a different genesis id, which is a different chain, and a different
/// one again after each restart.
#[derive(Clone, Debug, Default, PartialEq, Eq)]
pub struct Genesis {
    pub timestamp: i64,
}

impl Genesis {
    /// Read a genesis file.
    ///
    /// `initialTransactions` is REFUSED rather than ignored. The genesis
    /// allocation is folded into the committed state root before block 1, so a
    /// chain that skipped it would derive a different root for every later
    /// block — a fork, reported as nothing at all. This port carries no JSON
    /// transaction reader, and says so instead of pretending the file was
    /// empty.
    pub fn parse(raw: &[u8]) -> Result<Genesis> {
        if raw.is_empty() {
            return Ok(Genesis::default());
        }
        let v: serde_json::Value =
            serde_json::from_slice(raw).map_err(|e| Error::Config(e.to_string()))?;
        if let Some(txs) = v.get("initialTransactions").and_then(|x| x.as_array()) {
            if !txs.is_empty() {
                return Err(Error::Config(
                    "genesis names initial transactions, which this port does not read".into(),
                ));
            }
        }
        Ok(Genesis {
            timestamp: v.get("timestamp").and_then(|x| x.as_i64()).unwrap_or(0),
        })
    }
}

/// The Z-chain.
pub struct Zvm {
    cfg: Config,
    /// `sha256(ChainID ‖ NetworkID)`, hashed into every id this chain derives
    /// and into every proof's public inputs.
    bind: Id,
    db: View,
    store: Store,
    root: Root,
    nullifiers: Nullifiers,
    utxos: Utxos,
    verifier: Verifier,
    pending: Vec<Tx>,
    genesis: Id,
}

impl Zvm {
    /// Stand a chain up over an empty keyspace, seeded with `genesis`.
    pub fn new(cfg: Config, chain_id: Id, network_id: u32, genesis: &[u8]) -> Result<Zvm> {
        let g = Genesis::parse(genesis)?;
        let bind = block::binding(&chain_id, network_id);
        let verifier = Verifier::new(&cfg, bind)?;

        let first = Block::new(
            &bind,
            crate::wire::BlockBody {
                parent: ids::EMPTY,
                height: 0,
                timestamp: g.timestamp,
                txs: Vec::new(),
                state_root: Vec::new(),
                proof: None,
            },
        );
        let genesis_id = first.id();

        let mut db = View::new(Memory::new());
        let (store, fresh) = Store::open(&db, &bind, first)?;
        let mut root = Root::open(&db)?;
        let nullifiers = Nullifiers::open(&db)?;
        let utxos = Utxos::open(&db)?;

        if fresh {
            // The genesis fold is the one mutation outside a block, and it is
            // committed on its own: staged, it would ride on whichever block
            // landed first and vanish from a chain that never accepted one.
            let seeded = root.after(&[]);
            root.finalize(&mut db, &seeded)?;
            db.commit();
        }

        Ok(Zvm {
            cfg,
            bind,
            db,
            store,
            root,
            nullifiers,
            utxos,
            verifier,
            pending: Vec::new(),
            genesis: genesis_id,
        })
    }

    /// The chain binding every id and every proof is checked under.
    pub fn binding(&self) -> Id {
        self.bind
    }

    /// The genesis block's id.
    pub fn genesis(&self) -> Id {
        self.genesis
    }

    pub fn last_accepted(&self) -> Id {
        self.store.tip().0
    }

    pub fn height(&self) -> u64 {
        self.store.tip().1
    }

    pub fn state_root(&self) -> &[u8] {
        self.root.get()
    }

    pub fn nullifier_count(&self) -> u64 {
        self.nullifiers.count()
    }

    pub fn output_count(&self) -> u64 {
        self.utxos.count()
    }

    /// The height a nullifier was spent at, or `None`.
    pub fn spent_at(&self, nullifier: &[u8]) -> Result<Option<u64>> {
        self.nullifiers.spent(&self.db, nullifier)
    }

    pub fn block_id_at(&self, height: u64) -> Result<Id> {
        Store::id_at_height(&self.db, height)
    }

    /// A block by id: one in flight, the tip, or one read back from committed
    /// state.
    pub fn block(&self, id: &Id) -> Result<Block> {
        match self.store.get(&self.db, &self.bind, id)? {
            Decided::Block(b) => Ok(b),
            Decided::Vertex(_) => Err(Error::NotABlock(*id)),
        }
    }

    /// Read a block off the wire. It must not need the parent: a bootstrapping
    /// node parses blocks whose parents it does not hold yet.
    pub fn parse_block(&self, raw: &[u8]) -> Result<Block> {
        Block::parse(&self.bind, raw)
    }

    /// Record what the engine wants the next block built on.
    pub fn prefer(&mut self, id: Id) {
        self.store.prefer(id);
    }

    /// Offer a transaction to the pool. The same predicate a block is held to.
    pub fn submit(&mut self, tx: Tx) -> Result<()> {
        let next = self.store.tip().1 + 1;
        self.admit(&tx, next)?;
        self.pending.push(tx);
        Ok(())
    }

    /// Assemble a block from whatever is pending, on the preferred parent.
    ///
    /// Assembly runs the SAME predicate verification runs and drops what it
    /// cannot build, so a proposer never produces a block its own peers refuse.
    pub fn build(&mut self) -> Result<Block> {
        let parent = self.store.parent().clone();
        let height = parent.height() + 1;
        let taken = std::mem::take(&mut self.pending);
        let mut txs = Vec::with_capacity(taken.len());
        for tx in taken {
            if self.admit(&tx, height).is_ok() {
                txs.push(tx);
            }
        }
        if txs.is_empty() {
            return Err(Error::NoTransactions);
        }
        txs.truncate(self.cfg.max_tx_per_block as usize);

        // Chain time only moves forward, and verification refuses a block below
        // its parent. A parent may legally stand up to the skew allowance ahead
        // of this node's clock, so an unclamped clock here builds a block this
        // node's own verification then refuses.
        let mut timestamp = now();
        if let Decided::Block(pb) = &parent {
            timestamp = timestamp.max(pb.timestamp());
        }

        let state_root = self.root.after(&txs);
        Ok(Block::new(
            &self.bind,
            crate::wire::BlockBody {
                parent: parent.id(),
                height,
                timestamp,
                txs,
                state_root,
                proof: None,
            },
        ))
    }

    /// The whole verdict on a block, in the reference's order.
    pub fn verify(&mut self, b: &Block) -> Result<()> {
        b.shape(self.cfg.max_tx_per_block, now())?;

        for tx in b.txs() {
            self.admit(tx, b.height())?;
        }
        if b.body().proof.is_some() {
            self.verifier.verify_block_proof(b.txs())?;
        }

        if b.height() > 0 {
            let parent = self.block(&b.parent())?;
            // The parent must be one this chain can still build on: the
            // accepted tip, or a block verified above it and not yet decided.
            // Height alone is not that check — a block whose parent is an OLD
            // accepted block satisfies height == parent+1 perfectly well, and
            // accepting it rewinds the tip and leaves the height index naming
            // an orphan as the block at that height to every peer that
            // bootstraps from it.
            let (tip, tip_height) = self.store.tip();
            if parent.id() != tip && parent.height() <= tip_height {
                return Err(Error::NotOnTip(format!(
                    "parent {} at height {} is beneath the tip at {}",
                    ids::hex(&parent.id()),
                    parent.height(),
                    tip_height
                )));
            }
            if b.height() != parent.height() + 1 {
                return Err(Error::InvalidHeight);
            }
            if b.timestamp() < parent.timestamp() {
                return Err(Error::InvalidTimestamp);
            }
        }

        if b.state_root() != self.root.after(b.txs()).as_slice() {
            return Err(Error::InvalidStateRoot);
        }

        // A block that verifies is one the engine may build on, so it has to be
        // findable by id — including one parsed from a peer rather than built
        // here. Tracking only self-built blocks leaves a follower able to
        // verify the first block of a run and unable to verify the second.
        self.store.track(Decided::Block(b.clone()));
        Ok(())
    }

    /// Commit a block: its record, its spends, its outputs and the advanced
    /// root, in one commit or none of them.
    pub fn accept(&mut self, b: &Block) -> Result<()> {
        let decided = Decided::Block(b.clone());
        let next = self.root.after(b.txs());

        if let Err(e) = self.stage(b, &decided, &next) {
            // The abort has to reach the caches those writes also touched, or
            // the chain holds in memory what the database does not hold.
            self.db.abort();
            self.reload()?;
            return Err(e);
        }
        self.db.commit();

        let mut accepted = b.clone();
        accepted.set_status(crate::host::Status::Accepted);
        self.store.advance(Decided::Block(accepted));
        self.pending.clear();
        Ok(())
    }

    fn stage(&mut self, b: &Block, decided: &Decided, next: &[u8]) -> Result<()> {
        Store::stage(&mut self.db, decided)?;
        for tx in b.txs() {
            for n in &tx.nullifiers {
                self.nullifiers.mark(&mut self.db, n, b.height())?;
            }
            let id = tx.id();
            for (i, out) in tx.outputs.iter().enumerate() {
                self.utxos.add(
                    &mut self.db,
                    &Utxo {
                        tx: id,
                        output: i as u32,
                        commitment: out.commitment.clone(),
                        ciphertext: out.note.clone(),
                        ephemeral: out.ephemeral.clone(),
                        height: b.height(),
                    },
                )?;
            }
        }
        self.root.finalize(&mut self.db, next)
    }

    /// Put the caches back to what the database now says.
    fn reload(&mut self) -> Result<()> {
        self.root.reload(&self.db)?;
        self.nullifiers.reload(&self.db)?;
        self.utxos.reload(&self.db)
    }

    /// Drop a block that lost. It never wrote anything, so there is nothing
    /// else to undo.
    pub fn reject(&mut self, id: &Id) {
        self.store.forget(id);
    }

    /// The ONE predicate: what a transaction says about itself, whether it is
    /// still live at `height`, and whether its proof is good for it.
    pub fn admit(&mut self, tx: &Tx, height: u64) -> Result<()> {
        tx.validate_basic()?;
        if tx.expiry < height {
            return Err(Error::Expired);
        }
        self.verify_tx(tx)
    }

    fn verify_tx(&mut self, tx: &Tx) -> Result<()> {
        // A read that FAILED refuses the transaction: reporting "not spent" for
        // a set that could not be read is how an already-spent note gets spent
        // again.
        for n in &tx.nullifiers {
            match self.nullifiers.spent(&self.db, n) {
                Err(e) => return Err(Error::SpentSetRead(Box::new(e))),
                Ok(Some(_)) => return Err(Error::NullifierSpent),
                Ok(None) => {}
            }
        }
        self.verifier
            .verify_tx(tx)
            .map_err(|e| Error::ProofVerification(Box::new(e)))
    }
}

/// This node's clock, in seconds. The reference reads the wall clock, so a
/// chain held to a stated time would refuse blocks the reference admits.
fn now() -> i64 {
    std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .map(|d| d.as_secs() as i64)
        .unwrap_or(0)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::txs::{Kind, Proof, Shielded};

    fn chain() -> Zvm {
        Zvm::new(
            Config::chain_default(),
            ids::repeated(40),
            1,
            br#"{"timestamp":1000}"#,
        )
        .expect("a Z-chain")
    }

    fn transfer(nullifier: u8) -> Tx {
        Tx {
            kind: Kind::TRANSFER,
            version: 1,
            nullifiers: vec![vec![nullifier; 32]],
            outputs: vec![Shielded {
                commitment: vec![0x50; 32],
                note: vec![0x51; 64],
                ephemeral: vec![0x52; 32],
                range_proof: vec![0x53; 64],
            }],
            proof: Some(Proof {
                system: b"stark".to_vec(),
                data: vec![0x60; 192],
                public: vec![vec![0x61; 32], vec![0x62; 32]],
            }),
            fee: 1,
            expiry: 1000,
            ..Tx::default()
        }
    }

    fn empty_block(vm: &Zvm, parent: Id, height: u64, timestamp: i64) -> Block {
        Block::new(
            &vm.bind,
            crate::wire::BlockBody {
                parent,
                height,
                timestamp,
                txs: Vec::new(),
                state_root: vm.root.after(&[]),
                proof: None,
            },
        )
    }

    /// The genesis fold happens before block 1, so a chain born from a genesis
    /// with no transactions still starts ONE fold in rather than at the zero
    /// root. A block that named the zero root would be refused on a chain that
    /// did the fold and accepted on one that did not.
    #[test]
    fn the_chain_starts_one_fold_in() {
        let vm = chain();
        assert_ne!(vm.state_root(), vec![0u8; 32]);
        assert_eq!(vm.state_root(), Root::open(&Memory::new()).unwrap().after(&[]));
    }

    #[test]
    fn an_empty_block_on_the_seeded_chain_verifies() {
        let mut vm = chain();
        let tip = vm.last_accepted();
        let b = empty_block(&vm, tip, 1, 1000);
        vm.verify(&b).expect("an empty block on the tip");
    }

    /// The same block but for the parent it names. One must verify and the
    /// other must not, and the difference has to be the ledger.
    #[test]
    fn a_parent_no_chain_holds_is_a_ledger_refusal() {
        let mut vm = chain();
        let b = empty_block(&vm, ids::repeated(1), 1, 1000);
        assert!(matches!(vm.verify(&b), Err(Error::NoBlock(_))));
    }

    #[test]
    fn a_height_zero_block_that_names_a_parent_is_two_claims() {
        let mut vm = chain();
        let tip = vm.last_accepted();
        let b = empty_block(&vm, tip, 0, 1000);
        assert_eq!(vm.verify(&b), Err(Error::InvalidBlock));
    }

    #[test]
    fn a_state_root_the_chain_did_not_compute_is_refused() {
        let mut vm = chain();
        let tip = vm.last_accepted();
        let b = Block::new(
            &vm.bind,
            crate::wire::BlockBody {
                parent: tip,
                height: 1,
                timestamp: 1000,
                txs: Vec::new(),
                state_root: vec![0xAB; 32],
                proof: None,
            },
        );
        assert_eq!(vm.verify(&b), Err(Error::InvalidStateRoot));
    }

    /// The corpus's proofs are 192 bytes of one repeated value, which is not a
    /// proof of the one system a strict-PQ chain runs. It is refused for what
    /// the verifier said, under the wrapper the reference puts on it.
    #[test]
    fn a_shielded_transaction_is_refused_by_its_proof_and_not_by_its_shape() {
        let mut vm = chain();
        let tip = vm.last_accepted();
        let txs = vec![transfer(0x10)];
        let b = Block::new(
            &vm.bind,
            crate::wire::BlockBody {
                parent: tip,
                height: 1,
                timestamp: 1000,
                state_root: vm.root.after(&txs),
                txs,
                proof: None,
            },
        );
        let e = vm.verify(&b).unwrap_err();
        assert!(matches!(e, Error::ProofVerification(_)), "{e}");
        assert!(!e.before_the_chain(), "a proof needed the verifier");
    }

    /// A classical system is refused for what it IS, before anything judges it.
    #[test]
    fn a_classical_proof_is_forbidden_by_the_profile() {
        let mut vm = chain();
        let mut tx = transfer(0x1a);
        tx.proof.as_mut().unwrap().system = b"groth16".to_vec();
        let e = vm.admit(&tx, 1).unwrap_err();
        assert_eq!(e.root(), &Error::StrictPqClassicalForbidden);
    }

    #[test]
    fn a_transaction_past_its_expiry_is_refused_at_the_height_that_carries_it() {
        let mut vm = chain();
        let mut tx = transfer(0x19);
        tx.expiry = 1;
        assert_eq!(vm.admit(&tx, 2), Err(Error::Expired));
    }

    /// Acceptance advances the tip, the spent set, the output set and the root
    /// together, and a later block builds on what it left.
    #[test]
    fn acceptance_moves_everything_at_once() {
        let mut vm = chain();
        let tip = vm.last_accepted();
        let one = empty_block(&vm, tip, 1, 1000);
        vm.verify(&one).expect("verify");
        vm.accept(&one).expect("accept");

        assert_eq!(vm.last_accepted(), one.id());
        assert_eq!(vm.height(), 1);
        assert_eq!(vm.block_id_at(1).unwrap(), one.id());
        assert_eq!(vm.block(&one.id()).unwrap().id(), one.id());

        let two = empty_block(&vm, one.id(), 2, 1000);
        vm.verify(&two).expect("a block on the new tip");
    }

    #[test]
    fn a_genesis_naming_transactions_this_port_cannot_read_is_refused() {
        let e = Genesis::parse(br#"{"timestamp":1,"initialTransactions":[{}]}"#).unwrap_err();
        assert!(matches!(e, Error::Config(_)), "{e}");
        // An empty list is the same chain as no list, so it is not a refusal.
        assert_eq!(
            Genesis::parse(br#"{"timestamp":7,"initialTransactions":[]}"#).unwrap(),
            Genesis { timestamp: 7 }
        );
    }
}

File: chains/cpp/zkvm/src/vm.cpp
// Copyright (C) 2026, Lux Industries Inc. All rights reserved.
// SPDX-License-Identifier: BSD-3-Clause-Eco

#include "lux/zkvm/vm.hpp"

#include <chrono>
#include <set>

namespace lux::zkvm {

Id chain_binding(const Id& chain_id, std::uint32_t network_id) {
    Hasher h;
    h.write(view(chain_id));
    h.num32(network_id);
    return h.sum();
}

Vm::Vm(VmConfig config, store::Store& base)
    : config_(std::move(config)), pool_(config_.mempool_size) {
    bind_ = chain_binding(config_.chain_id, config_.network_id);
    chain_ = std::make_unique<ChainStore>(base, [this] { return reload(); });
}

std::int64_t Vm::now() const {
    if (now_ != 0) return now_;
    return std::chrono::duration_cast<std::chrono::seconds>(
               std::chrono::system_clock::now().time_since_epoch())
        .count();
}

wire::Result<void> Vm::reload() {
    if (auto r = nullifier_db_->reload(); !r) return r;
    if (auto r = utxo_db_->reload(); !r) return r;
    return root_->reload();
}

wire::Result<void> Vm::initialize(const Genesis& genesis) {
    if (config_.z.proof_cache_size == 0) config_.z.proof_cache_size = kDefaultProofCacheSize;
    if (config_.z.max_tx_per_block == 0) config_.z.max_tx_per_block = kDefaultMaxTxPerBlock;

    // The three stores write through the chain store's view, so their records
    // commit with the decision that made them and are discarded with one that
    // fails.
    auto utxo = UtxoDb::open(chain_->state());
    if (!utxo) return std::unexpected("failed to initialize UTXO DB: " + utxo.error());
    utxo_db_ = std::move(*utxo);

    auto nullifiers = NullifierDb::open(chain_->state());
    if (!nullifiers)
        return std::unexpected("failed to initialize nullifier DB: " + nullifiers.error());
    nullifier_db_ = std::move(*nullifiers);

    auto root = Root::open(chain_->state());
    if (!root) return std::unexpected("failed to initialize state root: " + root.error());
    root_ = std::move(*root);

    auto verifier = ProofVerifier::open(config_.z, bind_);
    if (!verifier)
        return std::unexpected("failed to initialize proof verifier: " + verifier.error());
    verifier_ = std::move(*verifier);

    // The Z-Chain accepts user-submitted shielded transactions, so it declares
    // the floor; the boot-time gate refuses a zero-fee user-facing chain.
    fee_ = Fee::floor();
    if (auto r = fee_.validate(); !r) return std::unexpected("zkvm: fee policy: " + r.error());

    genesis_ = std::make_shared<Block>();
    genesis_->bind_vm(*this);
    genesis_->block_height = 0;
    genesis_->block_timestamp = genesis.timestamp;
    genesis_->txs = genesis.initial_txs;
    genesis_->set_id(genesis_->compute_id());

    auto fresh = chain_->open(std::static_pointer_cast<Decision>(genesis_),
                              [this](ByteView raw) -> wire::Result<std::shared_ptr<Decision>> {
                                  auto b = parse_block(raw);
                                  if (!b) return std::unexpected(b.error());
                                  return std::static_pointer_cast<Decision>(*b);
                              });
    if (!fresh) return std::unexpected(fresh.error());

    // The genesis allocation is the one mutation outside a block, and it is
    // committed on its own: staged, it would ride on whichever block landed
    // first and vanish from a chain that never accepted one. seed records the
    // tip in that same commit, so a chain that has allocated says so on the next
    // boot and is not asked to allocate again — which ended in "UTXO already
    // exists", a node unable to restart until it produced a block.
    if (*fresh) {
        if (auto r = chain_->seed([&](store::Store& view_store) {
                return apply_genesis(genesis, view_store);
            });
            !r)
            return r;
    }
    return {};
}

wire::Result<void> Vm::apply_genesis(const Genesis& genesis, store::Store& view_store) {
    (void)view_store;
    for (const auto& tx : genesis.initial_txs) {
        for (std::size_t i = 0; i < tx.outputs.size(); ++i) {
            Utxo u;
            u.tx_id = tx.id;
            u.output_index = std::uint32_t(i);
            u.commitment = tx.outputs[i].commitment;
            u.ciphertext = tx.outputs[i].encrypted_note;
            u.ephemeral_pk = tx.outputs[i].ephemeral_pubkey;
            u.height = 0;  // genesis height
            if (auto r = utxo_db_->add(u); !r) return r;
        }
    }
    // Commit genesis to the root so block 1 builds on it, rather than leaving
    // genesis to be re-folded into every later block's root.
    return root_->finalize(root_->after(genesis.initial_txs));
}

Id Vm::compute_state_root(const std::vector<Transaction>& txs) const {
    return root_->after(txs);
}

wire::Result<void> Vm::admit(const Transaction& tx, std::uint64_t height) const {
    if (auto r = syntactic_verify(tx, height); !r) return r;
    return verify_transaction(tx);
}

wire::Result<void> Vm::syntactic_verify(const Transaction& tx, std::uint64_t height) const {
    if (auto r = tx.validate_basic(); !r) return r;
    if (tx.expiry < height) return std::unexpected(kErrExpired);
    return {};
}

wire::Result<void> Vm::verify_transaction(const Transaction& tx) const {
    for (const auto& n : tx.nullifiers) {
        auto spend = nullifier_db_->spent_at(view(n));
        if (!spend) return std::unexpected("zkvm: read spent set: " + spend.error());
        if (spend->spent) return std::unexpected(kErrNullifierSpent);
    }
    if (auto r = verifier_->verify(tx); !r)
        return std::unexpected("proof verification failed: " + r.error());
    return {};
}

wire::Result<Id> Vm::issue(Transaction tx) {
    if (auto r = fee_.admit(tx.fee); !r) return std::unexpected(r.error());
    if (auto r = tx.validate_basic(); !r) return std::unexpected(r.error());
    const Id id = tx.compute_id();
    if (auto r = pool_.add(std::move(tx)); !r) return std::unexpected(r.error());
    // The pool derives the id from the content, so this reports the transaction
    // the chain will carry rather than the one the caller named.
    return id;
}

wire::Result<std::shared_ptr<Block>> Vm::parse_block(ByteView raw) {
    auto blk = std::make_shared<Block>();
    blk->bind_vm(*this);
    if (auto r = parse_block_bytes(raw, *blk); !r) return std::unexpected(r.error());
    blk->set_bytes(bytes_of(raw));
    blk->set_id(blk->compute_id());
    return blk;
}

// block resolves a block id to a block. It is the ONE place this chain decides
// that a decision it made is a block: a vertex is a decision too, and asking for
// one by block id is answered as a miss rather than with something the caller
// cannot use.
wire::Result<std::shared_ptr<Block>> Vm::block(const Id& id) const {
    if (genesis_ && id == genesis_->id()) return genesis_;
    auto d = chain_->block(id, [this](ByteView raw) -> wire::Result<std::shared_ptr<Decision>> {
        auto b = const_cast<Vm*>(this)->parse_block(raw);
        if (!b) return std::unexpected(b.error());
        return std::static_pointer_cast<Decision>(*b);
    });
    if (!d) return std::unexpected(d.error());
    auto blk = std::dynamic_pointer_cast<Block>(*d);
    if (!blk) return std::unexpected(std::string(kErrNoBlock) + ": " + hex(id) + " is not a block");
    return blk;
}

wire::Result<std::shared_ptr<Block>> Vm::build_block() {
    auto built = chain_->propose(
        [&](const Decision& parent) -> wire::Result<std::shared_ptr<Decision>> {
            auto txs = pool_.pending(config_.z.max_tx_per_block);
            if (txs.empty()) return std::unexpected(kErrNoTransactions);

            // Assembly runs the SAME predicate verify runs, and drops what it
            // cannot build. It used to skip the shape check, so a transaction
            // with an out-of-range type — which the parser reads straight off
            // the wire — was assembled into every block and then refused by
            // every node's verify, including the proposer's. Nothing evicted it,
            // so that proposer never produced another block.
            std::vector<Transaction> valid;
            valid.reserve(txs.size());
            for (const auto& tx : txs) {
                if (auto r = admit(tx, parent.height() + 1); !r) {
                    pool_.remove(tx.id);
                    continue;
                }
                valid.push_back(tx);
            }
            if (valid.empty()) return std::unexpected(kErrNoTransactions);

            // Chain time only moves forward, and verify refuses a block below
            // its parent. A parent may legally be up to the clock skew ahead of
            // this node's clock, so an unclamped clock read here builds a block
            // this node's own verify then refuses.
            std::int64_t timestamp = now();
            const auto* parent_block = dynamic_cast<const Block*>(&parent);
            if (parent_block != nullptr && timestamp < parent_block->block_timestamp)
                timestamp = parent_block->block_timestamp;

            auto blk = std::make_shared<Block>();
            blk->bind_vm(*this);
            blk->parent_id = parent.id();
            blk->block_height = parent.height() + 1;
            blk->block_timestamp = timestamp;
            blk->txs = std::move(valid);
            const Id root = compute_state_root(blk->txs);
            blk->state_root.assign(root.begin(), root.end());
            blk->set_id(blk->compute_id());
            return std::static_pointer_cast<Decision>(blk);
        });
    if (!built) return std::unexpected(built.error());
    return std::static_pointer_cast<Block>(*built);
}

// build_vertex drains the pool, batches non-conflicting transactions and returns
// a vertex. admit is the same predicate verify runs, so nothing is batched that
// a peer will refuse.
wire::Result<std::shared_ptr<Vertex>> Vm::build_vertex() {
    const Id parent = chain_->tip();
    const std::uint64_t height = chain_->height();

    auto candidates = pool_.pending(config_.z.max_tx_per_block);
    if (candidates.empty()) return std::unexpected(kErrNoTransactions);

    ByteSet used;
    std::vector<Transaction> batch;
    for (const auto& tx : candidates) {
        if (auto r = admit(tx, height + 1); !r) {
            pool_.remove(tx.id);
            continue;
        }
        bool conflict = false;
        for (const auto& n : tx.nullifiers) {
            if (used.count(n)) {
                conflict = true;
                break;
            }
        }
        if (conflict) continue;
        for (const auto& n : tx.nullifiers) used.insert(n);
        batch.push_back(tx);
    }
    if (batch.empty()) return std::unexpected(kErrNoTransactions);

    auto v = std::make_shared<Vertex>();
    v->bind_vm(*this);
    v->vertex_height = height + 1;
    v->epoch = 0;
    v->parents = {parent};
    v->txs = std::move(batch);
    v->finish();
    return v;
}

wire::Result<std::shared_ptr<Vertex>> Vm::parse_vertex(ByteView raw) {
    return deserialize_vertex(raw, *this);
}

Health Vm::health() const {
    Health h;
    h.healthy = true;
    h.utxo_count = utxo_db_->count();
    h.nullifier_count = nullifier_db_->count();
    h.last_block_height = chain_->height();
    h.mempool_size = pool_.size();
    h.proof_cache_size = verifier_->cache_size();
    return h;
}

// ---- the node's seam ----

std::shared_ptr<lux::node::Block> Vm::build() {
    auto b = build_block();
    if (!b) return nullptr;  // nothing to build is "no", not a failure
    return std::static_pointer_cast<lux::node::Block>(*b);
}

std::shared_ptr<lux::node::Block> Vm::parse(std::span<const std::uint8_t> raw) {
    auto b = parse_block(raw);
    if (!b) return nullptr;
    return std::static_pointer_cast<lux::node::Block>(*b);
}

std::shared_ptr<lux::node::Block> Vm::get(const lux::node::Id& id) const {
    auto b = block(id);
    if (!b) return nullptr;
    return std::static_pointer_cast<lux::node::Block>(*b);
}

}  // namespace lux::zkvm
